src/hotspot/cpu/s390/foreignGlobals_s390.cpp

```diff
@@ -23,34 +23,210 @@
-  Unimplemented();
+  if (reg.type() == StorageType::INTEGER) {
+    __ reg2mem_opt(as_Register(reg), Address(Z_SP, offset), true);
+  } else if (reg.type() == StorageType::FLOAT) {
+    __ freg2mem_opt(as_FloatRegister(reg), Address(Z_SP, offset), true);
+  } else {
+    // stack and BAD
+  }
 }
 
 void RegSpiller::pd_load_reg(MacroAssembler* masm, int offset, VMStorage reg) {
-  Unimplemented();
+  if (reg.type() == StorageType::INTEGER) {
+    __ mem2reg_opt(as_Register(reg), Address(Z_SP, offset), true);
+  } else if (reg.type() == StorageType::FLOAT) {
+    __ mem2freg_opt(as_FloatRegister(reg), Address(Z_SP, offset), true);
+  } else {
+    // stack and BAD
+  }
+}
+
+static int reg2offset(VMStorage vms, int stk_bias) {
+  assert(!vms.is_reg(), "wrong usage");
+  return vms.index_or_offset() + stk_bias;
+}
+
+static void move_reg(MacroAssembler* masm, int out_stk_bias,
+                       VMStorage from_reg, VMStorage to_reg) {
+  int out_bias = 0;
+  switch (to_reg.type()) {
+    case StorageType::INTEGER:
+      if (to_reg.segment_mask() == REG64_MASK && from_reg.segment_mask() == REG32_MASK ) {
+        // see CCallingConventionRequiresIntsAsLongs
+        __ z_lgfr(as_Register(to_reg), as_Register(from_reg));
+      } else {
+        __ lgr_if_needed(as_Register(to_reg), as_Register(from_reg));
+      }
+      break;
+    case StorageType::STACK:
+      out_bias = out_stk_bias;  //fallthrough
+    case StorageType::FRAME_DATA: {
+      // Integer types always get a 64 bit slot in C.
+      if (from_reg.segment_mask() == REG32_MASK) {
+        // see CCallingConventionRequiresIntsAsLongs
+        __ z_lgfr(as_Register(from_reg), as_Register(from_reg));
+      }
+      switch (to_reg.stack_size()) {
+        case 8: __ reg2mem_opt(as_Register(from_reg), Address(Z_SP, reg2offset(to_reg, out_bias)), true);break;
+        case 4: __ reg2mem_opt(as_Register(from_reg), Address(Z_SP, reg2offset(to_reg, out_bias)), false);break;
+        default: ShouldNotReachHere();
+      }
+    }break;
+    default: ShouldNotReachHere();
+  }
+}
+
+static void move_float(MacroAssembler* masm, int out_stk_bias,
+                       VMStorage from_reg, VMStorage to_reg) {
+  switch (to_reg.type()) {
+    case StorageType::FLOAT:
+      if (from_reg.segment_mask() == REG64_MASK)
+        __ move_freg_if_needed(as_FloatRegister(to_reg), T_DOUBLE, as_FloatRegister(from_reg), T_DOUBLE);
+      else
+        __ move_freg_if_needed(as_FloatRegister(to_reg), T_FLOAT, as_FloatRegister(from_reg), T_FLOAT);
+      break;
+    case StorageType::STACK:
+      if (from_reg.segment_mask() == REG64_MASK) {
+        assert(to_reg.stack_size() == 8, "size should match");
+        __ freg2mem_opt(as_FloatRegister(from_reg), Address(Z_SP, reg2offset(to_reg, out_stk_bias)), true);
+      } else {
+        assert(to_reg.stack_size() == 4, "size should match");
+        // s390 needs 4 Byte offset
```

Comment:
Seems like this comment should get removed.

Reply:
Fixed